# DS 3000 HW 6

Due: Friday Nov 4 @ 11:59 PM EST

### Submission Instructions
Submit this `ipynb` file to Gradescope ([1:35 Section](https://www.gradescope.com/courses/431947) [3:25 Section](https://www.gradescope.com/courses/431864)) (this can also be done via the assignment on Canvas).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh `Kernel > Restart & Run All` just before uploading the `ipynb` file to gradescope.

### Tips for success
- Start early
- Make use of Piazza ([1:35 Section](https://piazza.com/class/l7dkfe3cgz21ul) [3:25 Section](https://piazza.com/class/l7dkey4g6m71e5)) (also accessable through Canvas)
- Make use of [Office Hours](https://docs.google.com/spreadsheets/d/1SMacPh9lhRvjHI1VCJfBbNGSga3RTYKzSsfRWEWqWQI/edit?usp=sharing)
- Remember to use cells and headings to make the notebook easy to read (if a grader cannot find the answer to a problem, you will receive no points for it)
    - Remember that documentation/style is important (See the Style Guide on Canvas!)
- Under no circumstances may one student view or share their ungraded homework or quiz with another student [(see also)](http://www.northeastern.edu/osccr/academic-integrity), though you are welcome to **talk about** (not show each other) the problems.

In [1]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

## Part 1 (18 points)

Match each of the covariance matrices to a single plot.  Note that all scatterplots below have equal scales in the x and y axes (moving an inch to the right increases x as much as moving an inch up increases y).  

Please express your response as list **in a markdown cell**, ordered alphabetically by dataset.  For example, you might respond with:

- Data Set A = $\Sigma_0$
- Data Set B = $\Sigma_1$
- Data Set C = $\Sigma_2$
- Data Set D = ...

(If you'r eunsure how to format a "Sigma" in Latex, double click this markdown cell and examine the source which generated the list above)

<img src="https://i.ibb.co/K7DY3mx/cov3by3.png" alt="Drawing" style="width: 600px;"/>


$$\Sigma_0 =\begin{bmatrix} 
5 & 1\\ 
1 & 2
\end{bmatrix}, 
\Sigma_1 =\begin{bmatrix} 
4 & 0\\ 
0 & 0
\end{bmatrix}, 
\Sigma_2 =\begin{bmatrix} 
1 & -1\\ 
-1 & 7
\end{bmatrix}, 
\Sigma_3 =\begin{bmatrix} 
7 & 4\\ 
4 & 5
\end{bmatrix}, 
\Sigma_4 =\begin{bmatrix} 
5 & 3\\ 
3 & 2
\end{bmatrix}$$

$$\Sigma_5 =\begin{bmatrix} 
6 & 2\\ 
2 & 6
\end{bmatrix}, 
\Sigma_6 =\begin{bmatrix} 
4 & -2\\ 
-2 & 5
\end{bmatrix}, 
\Sigma_7 =\begin{bmatrix} 
1 & -2\\ 
-2 & 5
\end{bmatrix}, 
\Sigma_8 =\begin{bmatrix} 
2 & 0\\ 
0 & 2
\end{bmatrix}$$

## Part 2: Electric Vehicles, Scale in Covariance Matrices (12 points)

This problem uses `evdataset.csv`, available in the Homework Module on Canvas, which was taken and adapted from [Kaggle](https://www.kaggle.com/datasets/mohamedalishiha/electric-vehicles) and contains a sample of 194 electric vehicles on the market until 2022. The full dataset includes basic technical specifications, battery capacity and range in various weather and road conditions. The adapted data set has fewer columns (see below).

In [2]:
import pandas as pd

df_ev = pd.read_csv('evdataset.csv', index_col='id')

In [3]:
df_ev.cov()

Acceleration    Top Speed  Electric Range   Total Power  \
Acceleration         10.038666  -105.351875     -227.482493   -325.912451   
Top Speed          -105.351875  1452.388361     2730.560734   4320.033278   
Electric Range     -227.482493  2730.560734    10278.791865   7944.660809   
Total Power        -325.912451  4320.033278     7944.660809  15114.115913   
Total Torque       -552.506025  7024.237968    13834.925218  24749.795577   
Battery Capacity    -36.213742   509.695970     1760.754834   1728.165045   
Charge Speed        -20.048900   169.335799      657.146921    355.204743   
Length               11.399997  3525.397602     8429.326024  16114.824155   
Width               -48.069676  1153.962769     1936.672988   4449.791144   
Height              356.311012 -3922.751910    -8336.923375  -9979.290316   
Wheelbase           180.853007   -58.100983     1480.904466   3358.385236   
Gross Weight        -71.458066  3976.754313    11345.072512  20151.268415   

                  Total Torque  Battery Capacity  Charge Speed         Length  \
Acceleration       -552.506025        -36.213742    -20.048900      11.399997   
Top Speed          7024.237968        509.695970    169.335799    3525.397602   
Electric Range    13834.925218       1760.754834    657.146921    8429.326024   
Total Power       24749.795577       1728.165045    355.204743   16114.824155   
Total Torque      49382.359810       3191.410138    459.734790   29169.851076   
Battery Capacity   3191.410138        404.922219     31.628682    4294.410323   
Charge Speed        459.734790         31.628682    243.919369   -2417.164494   
Length            29169.851076       4294.410323  -2417.164494  143049.769483   
Width              7577.200577        844.176625   -350.993590   22326.725122   
Height           -14288.262326       -661.694928  -1347.022675   21157.559559   
Wheelbase          8990.485978       2026.470757  -1677.429544   81208.268869   
Gross Weight      40044.338604       5491.262173  -2293.604749  142659.081005   

                         Width        Height     Wheelbase   Gross Weight  
Acceleration        -48.069676    356.311012    180.853007     -71.458066  
Top Speed          1153.962769  -3922.751910    -58.100983    3976.754313  
Electric Range     1936.672988  -8336.923375   1480.904466   11345.072512  
Total Power        4449.791144  -9979.290316   3358.385236   20151.268415  
Total Torque       7577.200577 -14288.262326   8990.485978   40044.338604  
Battery Capacity    844.176625   -661.694928   2026.470757    5491.262173  
Charge Speed       -350.993590  -1347.022675  -1677.429544   -2293.604749  
Length            22326.725122  21157.559559  81208.268869  142659.081005  
Width              4722.348486   2220.111533  12041.720421   24388.416217  
Height             2220.111533  23179.887319  20925.074115   24765.621414  
Wheelbase         12041.720421  20925.074115  56914.286977   85091.754367  
Gross Weight      24388.416217  24765.621414  85091.754367  172716.918354

A younger version of yourself, before having mastered Correlation and Covariance as you now have, may have had the following reasoning:

    "The covariance between Acceleration and Top Speed is -105.35, which is much closer to 0 than the covariance between Acceleration and Total Torque (-552.5).  Therefore, I conclude that Acceleration and Total Torque must change more consistently together than Acceleration and Top Speed."

Identify and correct your past self's interpretation mistake by offering a more principled analysis (**in a markdown cell**) of what relationships do and don't exist in the data.

To receive full credit on this question, be sure to:
- correctly identify the thinking error:
    - why is |cov(Acceleration, Total Torque)| > |cov(Acceleration, Top Speed)|? 
        - (use might use your `.head()` to do some thinking!)
        - also, notice that we are talking about *negative* covariance, so the absolute value (distance from zero) is what we are currently using to assess strength of relationship
    - what should we use to measure the consistency with which values change?
- clearly explain / teach the difference so that a non-technical audience may understand
- utilize whatever features of the python libraries we've studied you see fit to explain yourself

Be clear.  Be brief.

## Part 3: Electric Vehicles, k-NN Classifier

In this problem we'll use the k-Nearest Neighbor classifier to classify the type of drive an electric vehicle has (Rear, Front, or All-Wheel Drive (AWD)).

Note:
- each subpart requires the completion of all previous sub-parts
- the earlier subparts make scale normalization / cross validation errors on purpose
    - these will be addressed in later parts
- you may find it helpful to load a fresh copy of `df_ev` in parts 3.1, 3.3 and 3.4

## Part 3.1 (15 points)
Build a 20-Nearest Neighbor classifier which is `fit` and then `predict` on the whole dataset.

In [4]:
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib
import numpy as np

from sklearn.neighbors import KNeighborsClassifier

from sklearn.metrics import confusion_matrix
from sklearn.metrics import ConfusionMatrixDisplay

## Part 3.2 (15 points)
Plot the confusion matrix which results from this dataset.  

Using the confusion matrix, identify the drive mostly commonly misclassified by
- explicitly computing its accuracy
- telling which other drive it is most commonly mistaken for

Your response should be one or two sentences.

## Part 3.3 (15 points)
In a few sentences (**in a markdown cell**), describe how the scale of the data may be causing issues in the classifier's performance.  (Hint: see class notes on scale normalization).  Explicitly compute the variance of the raw `df_ev` dataset and refer to it in your explanation.

**THEN:** Repeat (i.e. copy/paste) parts 3.1 and 3.2, but **add in a fix** for this scaling problem.

## Part 3.4 (15 points)
In a few sentences, describe how the confusion matrix shown in part 3.3 may overstate how accurate the classifier would be in predicting the drive type of some new electric vehicle.  (Hint: see class notes on cross validation).  Update the code of part 3.3 to resolve this issue.

In [5]:
from sklearn.model_selection import StratifiedKFold

## Part 3.5 (10 points)

We will construct a graph that will help you choose the optimal $k$ based on the `accuracy_score` of the k-NN algorithm as applied to the electric vehicles data set. In this part, make this plot of the **cross validated** accuracy of the k-NN classifier for k = 1 to 50. **When you are done:** attach the graph as `best_k_ev.png` file with your Gradescope submission.

- in your cross validation, use `n_splits=10` folds of data
- write a function `get_cv_acc_knn()` which:
    - accepts:
        - `x`, `y_true`, `k` (of k-NN)
        - `n_splits=10` (defaults)
    - returns 
        - the cross validated accuracy of k-NN on the dataset
        
To accomplish this task, you will need the following two tools:

### Computing the `accuracy_score`

This is a single measure the tells you how well the classifier performed. Below is a small example where there are three groups (0, 1, 2):

In [6]:
from sklearn.metrics import accuracy_score

y_true_ex = [0, 0, 0, 0, 1, 2, 1, 0, 2, 0]
y_pred_ex = [1, 1, 1, 2, 1, 2, 0, 1, 2, 1]

accuracy_score(y_true_ex, y_pred_ex)

0.3

### Initializing empty arrays of specific size/type

In order to fill in your `y_pred` within the function, you will want to initialize an empty array of the same size and datatype as `y_true`, then loop through the `kfold.split(x, y_true)` to fill it in. Below is a small example:

In [7]:
y_true_ex = np.array(['a', 'b', 'c', 'd'])
y_pred_ex = np.empty_like(y_true_ex)
y_pred_ex

array(['', '', '', ''], dtype='<U1')